SimCalorimetry/HGCalSimProducers/src/HGCDigitizer.cc

```diff
@@ -13,13 +13,117 @@
 #include "FWCore/Framework/interface/Event.h"
 #include "FWCore/Framework/interface/EventSetup.h"
 #include "Geometry/Records/interface/IdealGeometryRecord.h"
+#include "Geometry/CaloGeometry/interface/CaloGeometry.h"
+#include "Geometry/Records/interface/CaloGeometryRecord.h"
 #include "Geometry/HGCalCommonData/interface/HGCalGeometryMode.h"
+#include "SimDataFormats/CaloTest/interface/HcalTestNumbering.h"
 
 #include <algorithm>
 #include <boost/foreach.hpp>
 
 using namespace hgc_digi;
 
+namespace {
+
+  float getPositionDistance(const HGCalGeometry* geom, const DetId& id) {
+    return geom->getPosition(id).mag();
+  }
+
+  float getPositionDistance(const HcalGeometry* geom, const DetId& id) {
+    return geom->getGeometry(id)->getPosition().mag();
+  }
+
+  void getValidDetIds(const HGCalGeometry* geom, std::vector<DetId>& valid) {
+    const std::vector<DetId>& ids = geom->getValidDetIds();
+    valid.reserve(ids.size());
+    valid.insert(valid.end(),ids.begin(),ids.end());
+  }
+
+  void getValidDetIds(const HcalGeometry* geom, std::vector<DetId>& valid) {
+    const std::vector<DetId>& ids = geom->getValidDetIds();
+    valid.reserve(ids.size());
+    for( const auto& id : ids ) {
+      if( DetId::Hcal == id.det() && 
+	  HcalEndcap == id.subdetId() ) 
+	valid.emplace_back(id);
+    }
+    valid.shrink_to_fit();    
+  }
+
+  DetId simToReco(const HcalGeometry* geom, unsigned simid) {
+    DetId result(0);
+    const auto& topo     = geom->topology();
```

Comment:
HcalGeometry eats an HcalTopology to be made and stores a pointer to the HcalTopology that made it.